addon/server.js

```diff
@@ -261,6 +287,104 @@ export default class Server {
   }
 
   /**
+   * Populates Mirage DB either from the persistence dump
+   * or by running the default scenario.
+   *
+   * @method loadScenario
+   * @public
+   */
+  loadScenario() {
+    if (!this.persistence) {
+      this.loadDefaultScenario();
+      return;
+    }
+
+    if (!window.localStorage) {
+      this.log("Mirage persistence is enabled, but localStorage is unavailable. Loading the default scenario.");
+      this.loadDefaultScenario();
+      return;
+    }
+
+    let dumpVersion = window.localStorage.getItem(LOCALSTORAGE_KEY_VERSION);
+
+    if (dumpVersion === undefined && this.persistenceVersion !== undefined) {
+      this.log("Mirage persistence dump not found. Loading the default scenario.");
+      this.loadDefaultScenario();
+      return;
+    }
+
+    if (dumpVersion !== undefined) {
+      dumpVersion = JSON.parse(dumpVersion);
+    }
+
+    if (this.persistenceVersion !== dumpVersion) {
+      this.log("Mirage persistence version mismatch. Loading the default scenario.");
+      window.localStorage.removeItem(LOCALSTORAGE_KEY_VERSION);
+      window.localStorage.removeItem(LOCALSTORAGE_KEY_DUMP);
+      this.loadDefaultScenario();
+      return;
+    }
+
+    let dump = window.localStorage.getItem(LOCALSTORAGE_KEY_DUMP);
+
+    if (!dump) {
+      this.log("Mirage persistence dump not found. Loading the default scenario.");
+      window.localStorage.removeItem(LOCALSTORAGE_KEY_VERSION);
+      this.loadDefaultScenario();
+      return;
+    }
+
+    let data = JSON.parse(dump);
+
+    this.log("Mirage persistence dump found, loading it into the database:", data);
+
+    this.db.loadData(data);
+  }
```

Comment:
I find the early returns in this method hard to read, could we move the validation logic into a method and have no early returns? It would also keep the style more consistent with other parts of the addon.

Something like

```js
loadScenario() {
  let dump = window.localStorage.getItem(LOCALSTORAGE_KEY_DUMP);
  let dumpVersionString = window.localStorage.getItem(LOCALSTORAGE_KEY_VERSION);
  let dumpVersion = (dumpVersion !== undefined) ? JSON.parse(dumpVersion) : dumpVersionString;

  if (this._shouldUseLocalStorageDump()) {
    let data = JSON.parse(dump);
    this.log("Mirage persistence dump found, loading it into the database:", data);
    this.db.loadData(data);
    
  } else {
    this.loadDefaultScenario();
  }
}

_shouldUseLocalStorageDump() {
  let shouldUseDump = true;
  
  if (!this.persistence) {
    shouldUseDump = false;
    
  } else if (!window.localStorage) {
    this.log("Mirage persistence is enabled, but localStorage is unavailable. Loading the default scenario.");
    shouldUseDump = false;
    
  } else if (dumpVersion === undefined && this.persistenceVersion !== undefined) {
    this.log("Mirage persistence dump not found. Loading the default scenario.");
    shouldUseDump = false;
    
  } else if (this.persistenceVersion !== dumpVersion) {
    this.log("Mirage persistence version mismatch. Loading the default scenario.");
    window.localStorage.removeItem(LOCALSTORAGE_KEY_VERSION);
    window.localStorage.removeItem(LOCALSTORAGE_KEY_DUMP);
    shouldUseDump = false;
    
  } else if (!dump) {
    this.log("Mirage persistence dump not found. Loading the default scenario.");
    window.localStorage.removeItem(LOCALSTORAGE_KEY_VERSION);
    shouldUseDump = false;
  }

  return shouldUseDump;
}
```